crates/sui-core/src/checkpoints2/casual_order.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::BTreeMap;
+use sui_types::base_types::TransactionDigest;
+use sui_types::messages::TransactionEffects;
+
+pub struct CasualOrder {
+    not_seen: BTreeMap<TransactionDigest, TransactionEffects>,
+    output: Vec<TransactionEffects>,
+}
+
+impl CasualOrder {
+    /// Casually sort given vector of effects
+    ///
+    /// Returned list has effects that
+    ///
+    /// (a) Casually sorted
+    /// (b) Have deterministic order between transactions that are not casually dependent
+    ///
+    /// The order of result list does not depend on order of effects in the supplied vector
+    pub fn casual_sort(effects: Vec<TransactionEffects>) -> Vec<TransactionEffects> {
+        let mut this = Self::from_vec(effects);
+        while let Some(item) = this.pop_first() {
+            this.insert(item);
+        }
+        this.into_list()
+    }
+
+    fn from_vec(effects: Vec<TransactionEffects>) -> Self {
+        let output = Vec::with_capacity(effects.len() * 2);
+        let not_seen = effects
+            .into_iter()
+            .map(|e| (e.transaction_digest, e))
+            .collect();
+        Self { not_seen, output }
+    }
+
+    fn pop_first(&mut self) -> Option<TransactionEffects> {
+        // Once map_first_last is stabilized this function can be rewritten as this:
+        // self.not_seen.pop_first()
+        let key = *self.not_seen.keys().next()?;
+        Some(self.not_seen.remove(key.as_ref()).unwrap())
+    }
+
+    // effect is already removed from self.not_seen at this point
+    fn insert(&mut self, effect: TransactionEffects) {
+        for dep in effect.dependencies.iter() {
+            if let Some(dep) = self.not_seen.remove(dep.as_ref()) {
+                self.insert(dep);
```

Comment:
How does this handle multiple layer of dependencies in the checkpoint? i.e. A depends on B which depends on C. And A is seen first.

Reply:
`self.insert` is a recursive call, that goes over dependencies first and outputs the node only after it's dependencies. The unit test for casual order actually have a case with 2 levels of dependencies